Unit Tests for AI Planning Subtrees not having required element/component

san7890 opened the pull request:
## About The Pull Request

Hey there,

I've personally fallen for this stupid thing twice (in #77503 and #75627 (d3575161ca4d73d3fa8fe44f59155572b4dcdfb2)), so I decided to spend a few hours to crack out a unit test to ensure that I (and no one else) falls for this stupid thing again.

Let me know if there's a smarter way to code something like this, but I couldn't figure out a better way to accomodate the current framework and be as agnostic to certain oddities as possible.
## Why It's Good For The Game
Catches stuff like this:

```txt
[2023-08-11 21:10:04.019]     FAILURE #1: The mob Garden Gnome does not have ANY instances of TRAIT_SUBTREE_REQUIRED_ELEMENT, but has a planning subtree (/datum/ai_planning_subtree/target_retaliate) that requires it! at code/modules/unit_tests/ensure_subtree_element.dm:45
 -     FAILURE #2: The mob the morph does not have ANY instances of TRAIT_SUBTREE_REQUIRED_ELEMENT, but has a planning subtree (/datum/ai_planning_subtree/target_retaliate) that requires it! at code/modules/unit_tests/ensure_subtree_element.dm:45
 -     FAILURE #3: The mob the guard spiderling (946) does not have ANY instances of TRAIT_SUBTREE_REQUIRED_ELEMENT, but has a planning subtree (/datum/ai_planning_subtree/target_retaliate/to_flee) that requires it! at code/modules/unit_tests/ensure_subtree_element.dm:45
 -     FAILURE #4: The mob the ambush spiderling (255) does not have ANY instances of TRAIT_SUBTREE_REQUIRED_ELEMENT, but has a planning subtree (/datum/ai_planning_subtree/target_retaliate/to_flee) that requires it! at code/modules/unit_tests/ensure_subtree_element.dm:45
 -     FAILURE #5: The mob the scout spiderling (375) does not have ANY instances of TRAIT_SUBTREE_REQUIRED_ELEMENT, but has a planning subtree (/datum/ai_planning_subtree/target_retaliate/to_flee) that requires it! at code/modules/unit_tests/ensure_subtree_element.dm:45
 -     FAILURE #6: The mob the flesh spiderling (337) does not have ANY instances of TRAIT_SUBTREE_REQUIRED_ELEMENT, but has a planning subtree (/datum/ai_planning_subtree/target_retaliate/to_flee) that requires it! at code/modules/unit_tests/ensure_subtree_element.dm:45
 -     FAILURE #7: The mob the hunter spiderling (869) does not have ANY instances of TRAIT_SUBTREE_REQUIRED_ELEMENT, but has a planning subtree (/datum/ai_planning_subtree/target_retaliate/to_flee) that requires it! at code/modules/unit_tests/ensure_subtree_element.dm:45
 -     FAILURE #8: The mob the nurse spiderling (629) does not have ANY instances of TRAIT_SUBTREE_REQUIRED_ELEMENT, but has a planning subtree (/datum/ai_planning_subtree/target_retaliate/to_flee) that requires it! at code/modules/unit_tests/ensure_subtree_element.dm:45
 -     FAILURE #9: The mob the tangle spiderling (19) does not have ANY instances of TRAIT_SUBTREE_REQUIRED_ELEMENT, but has a planning subtree (/datum/ai_planning_subtree/target_retaliate/to_flee) that requires it! at code/modules/unit_tests/ensure_subtree_element.dm:45
 -     FAILURE #10: The mob the broodmother spiderling (855) does not have ANY instances of TRAIT_SUBTREE_REQUIRED_ELEMENT, but has a planning subtree (/datum/ai_planning_subtree/target_retaliate/to_flee) that requires it! at code/modules/unit_tests/ensure_subtree_element.dm:45
 -     FAILURE #11: The mob the viper spiderling (519) does not have ANY instances of TRAIT_SUBTREE_REQUIRED_ELEMENT, but has a planning subtree (/datum/ai_planning_subtree/target_retaliate/to_flee) that requires it! at code/modules/unit_tests/ensure_subtree_element.dm:45
 -     FAILURE #12: The mob the tarantula spiderling (963) does not have ANY instances of TRAIT_SUBTREE_REQUIRED_ELEMENT, but has a planning subtree (/datum/ai_planning_subtree/target_retaliate/to_flee) that requires it! at code/modules/unit_tests/ensure_subtree_element.dm:45
 -     FAILURE #13: The mob the spiderling (100) does not have ANY instances of TRAIT_SUBTREE_REQUIRED_ELEMENT, but has a planning subtree (/datum/ai_planning_subtree/target_retaliate/to_flee) that requires it! at code/modules/unit_tests/ensure_subtree_element.dm:45
```

(ignore the part about gnomes and morphs, this was an earlier version of the unit test. everything else was relevant and is fixed)
## Changelog
:cl:
fix: Growing spiders will now retaliate against you like they were always meant to.
/:cl:

san7890 committed:
groundwork

san7890 committed:
man this is fugly

san7890 committed:
gwak gwak

san7890 committed:
screeeeee

san7890 committed:
this should be it

san7890 committed:
adds the type as well

san7890 committed:
fix up spiderlings at least

san7890 committed:
generalizes it to work for components too

san7890 committed:
fixes it for gnomes

san7890 committed:
never mind not done with gnomes yet

san7890 committed:
pain. agony. suffering

san7890 committed:
WHY CAN I NOT NEW SUBTREE AARRGHH

it's okay we still can get easy money

san7890 committed:
???

san7890 committed:
oh we don't need to access the ss

san7890 committed:
fix typo

ZephyrTFA commented on code/_compile_options.dm:
```diff
@@ -63,7 +63,7 @@
 
 // If this is uncommented, we do a single run though of the game setup and tear down process with unit tests in between
-// #define UNIT_TESTS
+#define UNIT_TESTS
```
```suggestion
// #define UNIT_TESTS
```

san7890 committed:
removes debug stuff

san7890 committed:
Merge remote-tracking branch 'upstream/master' into god-dammit

san7890 committed:
elaborates on a comment

san7890 committed:
fixes the string issue

san7890 committed:
code cleanup

san7890 committed:
adds it to `basic_ranged_attack_subtree` as well

san7890 committed:
alphabetizes the list

since we aren't accessing the index of the list and instead using the key this is better
